dkg/sync/server.go

```diff
@@ -28,191 +29,224 @@ import (
-				return
-			}
+	size := int64(len(buf))
+	err = binary.Write(writer, binary.LittleEndian, size)
+	if err != nil {
+		return errors.Wrap(err, "read size")
+	}
 
-			buf := bufio.NewReader(s)
-			b := make([]byte, MsgSize)
-			// n is the number of bytes read from buffer, if n < MsgSize the other bytes will be 0
-			n, err := buf.Read(b)
-			if err != nil {
-				log.Error(ctx, "Read client msg from stream", err, z.Any("client", p2p.PeerName(pID)))
-				return
-			}
+	n, err := writer.Write(buf)
+	if err != nil {
+		return errors.Wrap(err, "write message")
+	} else if int64(n) != size {
+		return errors.New("unexpected message length")
+	}
 
-			// The first `n` bytes that are read are the most important
-			b = b[:n]
+	return nil
+}
 
-			msg := new(pb.MsgSync)
-			if err := proto.Unmarshal(b, msg); err != nil {
-				log.Error(ctx, "Unmarshal client msg", err)
-				return
-			}
+// readSizedProto reads a size prefixed proto message.
+func readSizedProto(reader io.Reader, msg proto.Message) error {
+	var size int64
+	err := binary.Read(reader, binary.LittleEndian, &size)
+	if err != nil {
+		return errors.Wrap(err, "read size")
+	}
 
-			log.Debug(ctx, "Message received from client", z.Any("client", p2p.PeerName(pID)))
+	buf := make([]byte, size)
+	n, err := reader.Read(buf)
+	if err != nil {
+		return errors.Wrap(err, "read buffer")
+	} else if int64(n) != size {
+		return errors.New("unexpected message length")
+	}
 
-			if msg.Shutdown {
-				resp := &pb.MsgSyncResponse{
-					SyncTimestamp: msg.Timestamp,
-					Error:         "",
-				}
+	err = proto.Unmarshal(buf, msg)
+	if err != nil {
+		return errors.Wrap(err, "unmarshal proto")
+	}
 
-				resBytes, err := proto.Marshal(resp)
-				if err != nil {
-					log.Error(ctx, "Marshal server response", err)
-					return
-				}
+	return nil
+}
 
-				_, err = s.Write(resBytes)
-				if err != nil {
-					log.Error(ctx, "Send response to client", err, z.Any("client", p2p.PeerName(pID)))
-					return
-				}
-				server.receiveShutdown <- result{shutdown: true}
+func (s *Server) handleStream(ctx context.Context, stream network.Stream) error {
+	defer stream.Close()
 
-				continue
-			}
+	pID := stream.Conn().RemotePeer()
+	pubkey, err := pID.ExtractPublicKey()
+	if err != nil {
+		return errors.Wrap(err, "extract pubkey")
+	}
 
-			pubkey, err := pID.ExtractPublicKey()
-			if err != nil {
-				log.Error(ctx, "Get client public key", err)
-				return
-			}
+	defer s.clearConnected(pID)
 
-			ok, err := pubkey.Verify(defHash, msg.HashSignature)
-			if err != nil {
-				log.Error(ctx, "Verify defHash signature", err)
-				return
-			}
+	for {
+		// Read next sync message
+		msg := new(pb.MsgSync)
+		if err := readSizedProto(stream, msg); err != nil {
+			return err
+		}
 
-			resp := &pb.MsgSyncResponse{
-				SyncTimestamp: msg.Timestamp,
-				Error:         "",
-			}
+		if msg.Shutdown {
+			s.setShutdown(pID)
+		}
 
-			if !ok {
-				resp.Error = InvalidSig
-			}
+		// Verify definition hash
 
```

Comment:
remove this blank line

Reply:
I'm putting a blank line since "verify definition hash" is multiple steps, not just the next line.

Reply:
Ok cool, removed the blank line